File: Linked List/0234. Palindrome Linked List/0234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        ListNode* slow = head;
        ListNode* fast = head;

        while (fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }

        ListNode* prev = NULL;
        ListNode* curr = head;
        while (curr != slow) {
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        if (fast) { // odd case
            slow = slow->next;
        }

        while (slow) {
            if (slow->val != prev->val) {
                return false;
            } else {
                slow = slow->next;
                prev = prev->next;
            }
        }
        return true;
    }
};

File: Linked List/0234. Palindrome Linked List/0234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True
        
        slow,fast = head,head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        curr = head
        prev = None
        while curr != slow:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        if fast:
            slow = slow.next
        
        while slow:
            if slow.val != prev.val:
                return False
            else:
                slow = slow.next
                prev = prev.next
        return True